acvm/src/pwg/block.rs

```diff
@@ -0,0 +1,188 @@
+            let op_expr = ArithmeticSolver::evaluate(&block_op.operation, initial_witness);
+            let operation = expression_to_const(&op_expr).ok_or_else(|| {
+                OpcodeResolutionError::OpcodeNotSolvable(OpcodeNotSolvable::MissingAssignment(
+                    op_expr.any_witness().unwrap().0,
+                ))
+            })?;
+            let index_expr = ArithmeticSolver::evaluate(&block_op.index, initial_witness);
+            let index = expression_to_const(&index_expr).ok_or_else(|| {
+                OpcodeResolutionError::OpcodeNotSolvable(OpcodeNotSolvable::MissingAssignment(
+                    index_expr.any_witness().unwrap().0,
+                ))
+            })?;
+            let index = index.try_to_u64().unwrap() as u32;
+            let value = ArithmeticSolver::evaluate(&block_op.value, initial_witness);
+            let value_witness = value.any_witness();
+            if operation.is_zero() {
+                let value = ArithmeticSolver::evaluate(&block_op.value, initial_witness);
+                if value.is_const() {
+                    self.insert_value(index, value.q_c)?;
+                } else if value.is_linear() {
+                    match ArithmeticSolver::solve_fan_in_term(&value, initial_witness) {
+                        GateStatus::GateUnsolvable => {
+                            return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                                OpcodeNotSolvable::MissingAssignment(value_witness.unwrap().0),
+                            ))
+                        }
+                        GateStatus::GateSolvable(sum, (coef, w)) => {
+                            let map_value = self.get_value(index).ok_or(
+                                OpcodeResolutionError::OpcodeNotSolvable(
+                                    OpcodeNotSolvable::MissingAssignment(w.0),
+                                ),
+                            )?;
+                            insert_witness(
+                                w,
+                                (map_value - sum - value.q_c) / coef,
+                                initial_witness,
+                            )?;
+                        }
+                        GateStatus::GateSatisfied(sum) => {
+                            self.insert_value(index, sum + value.q_c)?;
+                        }
+                    }
+                } else {
+                    return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                        OpcodeNotSolvable::MissingAssignment(value_witness.unwrap().0),
+                    ));
+                }
+            } else if value.is_const() {
+                self.insert_value(index, value.q_c)?;
+            } else {
+                return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                    OpcodeNotSolvable::MissingAssignment(value_witness.unwrap().0),
+                ));
+            }
```

Comment:
```suggestion
            let value_witness_index = value.any_witness().unwrap().0;
            if value.is_const() {
                self.insert_value(index, value.q_c)?;
            } else if operation.is_zero() && value.is_linear() {
                match ArithmeticSolver::solve_fan_in_term(&value, initial_witness) {
                    GateStatus::GateUnsolvable => {
                        return Err(OpcodeResolutionError::OpcodeNotSolvable(
                            OpcodeNotSolvable::MissingAssignment(value_witness_index),
                        ))
                    }
                    GateStatus::GateSolvable(sum, (coef, w)) => {
                        let map_value = self.get_value(index).ok_or(
                            OpcodeResolutionError::OpcodeNotSolvable(
                                OpcodeNotSolvable::MissingAssignment(w.0),
                            ),
                        )?;
                        insert_witness(w, (map_value - sum - value.q_c) / coef, initial_witness)?;
                    }
                    GateStatus::GateSatisfied(sum) => {
                        self.insert_value(index, sum + value.q_c)?;
                    }
                }
            } else {
                return Err(OpcodeResolutionError::OpcodeNotSolvable(
                    OpcodeNotSolvable::MissingAssignment(value_witness_index),
                ));
            };
```

Couple of changes here:
- `value_witness` always gets unwrapped into its index so we can do this in one place (It would probably be cleaner to convert from `Witness` to index inside https://github.com/noir-lang/acvm/pull/114/files#r1117185097 but we should unwrap early if we know that this witness will always exist.)
- We're recalculating `value` here for no reason I can see.
- branching behaviour always results in the same outcome if `value.is_const()` so pull this to outside of the nested if-statements. This actually allows us to remove nesting altogether.

Reply:
Yes good points, thank you for the suggestion. It's updated.